### DISCUSSION TITLE
Running part of pipeline on SLURM cluster with future.batchtools over ssh

### DISCUSSION THREAD

**Discussion by edbonneville** [ropensci/targets#709] [Help] [answered]
Hi - thanks for the great work on the package! I read https://github.com/ropensci/targets/issues/198, https://github.com/ropensci/drake/issues/1295 and https://github.com/ropensci/drake/issues/906 while trying a particular setup for a `{targets}`-based simulation study.

As background: I am connecting to a SLURM cluster from a Windows machine, where reverse ssh tunneling with `{clustermq}` does not yet work (as far as I have tried). I do however have a working setup with `{future}` and `{future.batchtools}` - for example, running something like

```r
library("targets")
library("future")
library("future.batchtools")
library("furrr")

# Prepare plan
login <- future::tweak(future::remote, workers = "cluster_ip") 
slurm_run <- future::tweak(
  strategy = future.batchtools::batchtools_slurm,
  template = "~/future_slurm.tmpl",
  resources = list(
    walltime = 5,
    ntasks = 1,
    ncpus = 1,
    memory = '1G',
    partition = 'short'
  )
)

future::plan(list(login, slurm_run))
vec %<-% future_map(seq_len(4), ~ Sys.sleep(5))
vec
```
.. creates four jobs on the cluster, where the 'results' are then returned to my local session.

## `{targets}` question

Suppose I have the following setup:

```r
# _targets.R:
plan(sequential)
list(
  tar_target(x, seq_len(4)),
  tar_target(
    y,
    Sys.sleep(5),
    pattern = map(x),
    resources = tar_resources(
      future = tar_resources_future(plan = list(login, slurm_run))
    ),
    deployment = "worker"
  )
)

# Console:
tar_make_future(workers = 4)
```

Here, I am trying to run x 'sequentially' on my local machine, and then I would like a job for each of the four branches of y to be sent to the cluster via ssh (the branches would for example be the scenarios of a simulation study). The current setup does not seem to be sending anything to the cluster. Note that if I run the above directly on the cluster, using `tar_resources_future(plan = slurm_run)`, it works as expected.

Any ideas on how I can adjust the plan(s) to get the ssh part of the pipeline to work? Or maybe I need to change options elsewhere?

## Session Info

(Same version of R running on the cluster)

```r
- Session info ----------------------------------------------------------------------------
 setting  value                       
 version  R version 4.1.0 (2021-05-18)
 os       Windows 10 x64              
 system   x86_64, mingw32             
 ui       RStudio                     
 language (EN)                        
 collate  Dutch_Netherlands.1252      
 ctype    Dutch_Netherlands.1252      
 tz       Europe/Berlin               
 date     2021-11-19                  

- Packages --------------------------------------------------------------------------------
 package           * version date       lib source        
 backports           1.2.1   2020-12-09 [1] CRAN (R 4.0.3)
 base64url           1.4     2018-05-14 [1] CRAN (R 4.1.1)
 batchtools          0.9.15  2021-01-11 [1] CRAN (R 4.1.1)
 brew                1.0-6   2011-04-13 [1] CRAN (R 4.0.0)
 cachem              1.0.6   2021-08-19 [1] CRAN (R 4.1.1)
 callr               3.7.0   2021-04-20 [1] CRAN (R 4.0.5)
 checkmate           2.0.0   2020-02-06 [1] CRAN (R 4.0.2)
 cli                 3.0.0   2021-06-30 [1] CRAN (R 4.1.0)
 codetools           0.2-18  2020-11-04 [1] CRAN (R 4.0.3)
 crayon              1.4.2   2021-10-29 [1] CRAN (R 4.1.1)
 data.table          1.14.2  2021-09-27 [1] CRAN (R 4.1.1)
 desc                1.4.0   2021-09-28 [1] CRAN (R 4.1.1)
 devtools            2.4.2   2021-06-07 [1] CRAN (R 4.1.0)
 digest              0.6.27  2020-10-24 [1] CRAN (R 4.0.3)
 ellipsis            0.3.2   2021-04-29 [1] CRAN (R 4.0.5)
 fansi               0.5.0   2021-05-25 [1] CRAN (R 4.1.0)
 fastmap             1.1.0   2021-01-25 [1] CRAN (R 4.0.3)
 fs                  1.5.0   2020-07-31 [1] CRAN (R 4.0.2)
 furrr             * 0.2.3   2021-06-25 [1] CRAN (R 4.1.0)
 future            * 1.22.1  2021-08-25 [1] CRAN (R 4.1.1)
 future.batchtools * 0.10.0  2021-01-04 [1] CRAN (R 4.1.1)
 globals             0.14.0  2020-11-22 [1] CRAN (R 4.0.3)
 glue                1.4.2   2020-08-27 [1] CRAN (R 4.0.2)
 hms                 1.1.1   2021-09-26 [1] CRAN (R 4.1.1)
 igraph              1.2.6   2020-10-06 [1] CRAN (R 4.0.2)
 knitr               1.36    2021-09-29 [1] CRAN (R 4.1.1)
 lifecycle           1.0.1   2021-09-24 [1] CRAN (R 4.1.1)
 listenv             0.8.0   2019-12-05 [1] CRAN (R 4.0.2)
 magrittr            2.0.1   2020-11-17 [1] CRAN (R 4.0.3)
 memoise             2.0.0   2021-01-26 [1] CRAN (R 4.0.3)
 parallelly          1.28.1  2021-09-09 [1] CRAN (R 4.1.1)
 pillar              1.6.4   2021-10-18 [1] CRAN (R 4.1.1)
 pkgbuild            1.2.0   2020-12-15 [1] CRAN (R 4.0.3)
 pkgconfig           2.0.3   2019-09-22 [1] CRAN (R 4.0.2)
 pkgload             1.2.3   2021-10-13 [1] CRAN (R 4.1.1)
 prettyunits         1.1.1   2020-01-24 [1] CRAN (R 4.0.2)
 processx            3.5.2   2021-04-30 [1] CRAN (R 4.0.5)
 progress            1.2.2   2019-05-16 [1] CRAN (R 4.0.2)
 ps                  1.6.0   2021-02-28 [1] CRAN (R 4.0.4)
 purrr               0.3.4   2020-04-17 [1] CRAN (R 4.0.2)
 R6                  2.5.1   2021-08-19 [1] CRAN (R 4.1.1)
 rappdirs            0.3.3   2021-01-31 [1] CRAN (R 4.0.3)
 remotes             2.4.1   2021-09-29 [1] CRAN (R 4.1.1)
 rlang               0.4.11  2021-04-30 [1] CRAN (R 4.0.5)
 rprojroot           2.0.2   2020-11-15 [1] CRAN (R 4.0.3)
 sessioninfo         1.1.1   2018-11-05 [1] CRAN (R 4.0.2)
 stringi             1.6.1   2021-05-10 [1] CRAN (R 4.1.0)
 targets           * 0.8.1   2021-10-26 [1] CRAN (R 4.1.1)
 testthat            3.1.0   2021-10-04 [1] CRAN (R 4.1.1)
 tibble              3.1.2   2021-05-16 [1] CRAN (R 4.1.0)
 tidyselect          1.1.1   2021-04-30 [1] CRAN (R 4.0.5)
 usethis             2.1.3   2021-10-27 [1] CRAN (R 4.1.1)
 utf8                1.2.1   2021-03-12 [1] CRAN (R 4.0.4)
 vctrs               0.3.8   2021-04-29 [1] CRAN (R 4.0.5)
 withr               2.4.2   2021-04-18 [1] CRAN (R 4.0.5)
 xfun                0.27    2021-10-18 [1] CRAN (R 4.1.1)
 yaml                2.2.1   2020-02-01 [1] CRAN (R 4.0.2)
```

**Comment by wlandau** [comment, marked as answer]
That's a tough one, I don't think I can reproduce your connection setup.

In your first example, you write:

```r
vec %<-% future_map(seq_len(4), ~ Sys.sleep(5))
```

Here, you use both levels of the nested `future` plan: `%<-%` for the outer plan and `future_map()` for the inner plan. But in `tar_make_future()`, only the first level of the plan is used, so any inner plans are passed on to the custom user-side code of the target. If you were to write something like `tar_target(x, future(your_code())`, I predict the target would log in and then `your_code()` would run on SLURM.

So maybe run `tar_make_future()` inside the login plan and pass the SLURM plan to the targets? Might not work because `tar_make_future()` creates an internal process of its own unless `callr_function = NULL`, which should only be used for debugging and testing purposes. 

I was also thinking you might try `tar_resources_future(plan = future::tweak(list(login, slurm_run)))` instead of `tar_resources_future(plan = list(login, slurm_run))`, but that's probably not it.

**Reply by edbonneville** [reply]
Thanks a lot for quick reply - trying `tar_resources_future(plan = future::tweak(list(login, slurm_run)))` gives the error

```r
Error in UseMethod("tweak") : 
  no applicable method for 'tweak' applied to an object of class "list"
``` 
However using (in my example code) `future(Sys.sleep(5))`  did create the jobs on the cluster! I guess I hadn't quite wrapped my head around the plan levels in `{future}`. 

One more question for my understanding: when running `tar_make_future(workers = 4)` in this context, what does workers argument limit? Will it limit the maximum amount of SLURM jobs running at any given moment, or will all jobs be allocated at the same time (assuming there is space) even with `workers = 1`?

**Reply by wlandau** [reply]
`workers` caps the number of targets that `tar_make_future()` runs simultaneously at any given time. This could be different from the number of cores, futures, or SLURM jobs, depending on what is happening in the user's code. `workers = 1` means only one target runs at a time. It's like `tar_make()`, except each target is a SLURM job and the next target waits for the current job to finish.

**Comment by crossterra-agent** [comment]
- **Status**: BLOCKED
- **Role**: orchestrator
- **Link**: https://github.com/CrossTerra/monorepo/pull/2396
- **Summary**: All required CI checks pass on `f786046b`; Cursor Bugbot is still pending for that commit. No developer DONE report has been posted after the latest push. Continuing to wait — ci-watcher will invoke the developer if Bugbot fails; once the developer reports a successful DONE, I will invoke the reviewer.
